packages/core/src/external/mpi/match-patients.ts

```diff
@@ -0,0 +1,147 @@
+import { intersectionWith, isEqual } from "lodash";
+import { PatientDataMPI } from "./patient-incoming-schema";
+import { normalizePatientDataMPI } from "./normalize-patient";
+import jaroWinkler from "jaro-winkler";
+
+// Define a type for the similarity function
+type SimilarityFunction = (
+  patient1: PatientDataMPI,
+  patient2: PatientDataMPI,
+  threshold: number
+) => boolean;
+
+type MatchingRule = (patient1: PatientDataMPI, patient2: PatientDataMPI) => boolean;
+
+/**
+ * `matchPatients` filters patients based on a similarity function, rules, and a threshold.
+ * @param similarityFunction - Determines if a patient is a match.
+ * @param matchingRules - Rules that determines if a patient is a match.
+ * @param patients - Array of patients.
+ * @param demo - Patient data to find matches for.
+ * @param threshold - Minimum similarity score for a match.
+ * @param greedy - If true, returns the first match. If false, returns all matches.
+ * @returns matched patients.
+ */
+export const matchPatients = (
+  similarityFunction: SimilarityFunction,
+  matchingRules: MatchingRule[],
+  patients: PatientDataMPI[],
+  demo: PatientDataMPI,
+  threshold: number,
+  greedy = true
+): PatientDataMPI[] => {
+  const matchFunction = (patient: PatientDataMPI) => {
+    const PatientDataMPI = normalizePatientDataMPI(patient);
+    if (!PatientDataMPI) {
+      return false;
+    }
+    for (const rule of matchingRules) {
+      if (rule(demo, PatientDataMPI)) {
+        return true;
+      }
+    }
+    return similarityFunction(PatientDataMPI, demo, threshold);
+  };
+  if (greedy) {
+    const foundPatient = patients.find(matchFunction);
+    return foundPatient ? [foundPatient] : [];
+  } else {
+    return patients.filter(matchFunction);
+  }
+};
+
+/**
+ * This function checks if the patient has any personal identifiers that match the demo. The idea of having rules
+ * is that rules will override statistical similarity methods. This rule should act as an override.
+ * @param demo
+ * @param patient
+ * @returns true if the patient has any personal identifiers that match the demo.
+ */
+export const matchingPersonalIdentifiersRule = (
+  demo: PatientDataMPI,
+  patient: PatientDataMPI
+): boolean => {
+  return Boolean(
+    intersectionWith(patient.personalIdentifiers || [], demo.personalIdentifiers || [], isEqual)
+      .length > 0
+  );
+};
```

Comment:
Could be easier to understand if you check first if either of them is truthy and then combine the two if they are.

Reply:
@jonahkaye just in case you missed this

Reply:
The way I did it is the way lodash does it in the docs: https://lodash.com/docs/4.17.15#intersectionWith
I feel like its not too hard to understand, but can still change if others agree.

Reply:
raf also found it confusing so I guess was too close to the code on this one! changed it